packages/flutter/lib/src/material/mergeable_material.dart

```diff
@@ -12,10 +12,17 @@ import 'dart:ui' show lerpDouble;
 ///
 /// All [MergeableMaterialItem] objects need a [LocalKey].
 abstract class MergeableMaterialItem {
-  MergeableMaterialItem(this.key) {
-    assert(key != null);
-  }
-
+  /// Abstract const constructor. This constructor enables subclasses to provide
+  /// const constructors so that they can be used in const expressions.
+  ///
+  /// The argument is the [key], which must not be null.
+  const MergeableMaterialItem(this.key);
```

Comment:
It's going to require some trivial changes to the VM and analyzer setups first, but yes.